src/pykeen/utils.py

```diff
@@ -1095,3 +1100,72 @@ def load_configuration(path: Union[str, pathlib.Path, os.PathLike]) -> Mapping[s
     import doctest
 
     doctest.testmod()
+
+
+def product_normalise(input_tensor: torch.FloatTensor) -> torch.FloatTensor:
+    r"""Normalise the input tensor along its embedding dimension so that the geometric mean is 1.
+
+    :param input_tensor: An input tensor with final dimension $d$.
+    """
+    step1_tensor = torch.abs(input_tensor)  # Compute absolute value of all entries
+    step2_tensor = step1_tensor + SANITY_EPSILON  # Prevent zero values by adding a sanity epsilon
+    log_norm_tensor = torch.log(step2_tensor)  # Compute the log prior to computing the geom. mean
+    step3_tensor = torch.mean(log_norm_tensor, dim=-1, keepdim=True)
+    norm_volume = torch.exp(step3_tensor)
+    pre_norm_out = input_tensor / norm_volume
+    return pre_norm_out
+
+
+def compute_box(
+    base: torch.FloatTensor, delta: torch.FloatTensor, size: torch.FloatTensor
+) -> Tuple[torch.FloatTensor, torch.FloatTensor]:
+    r"""Given sets of embeddings of base position, shape, and size, compute the lower and upper corners of
+    the resulting box
+
+    :param base: The base position (box center) of the input relation embeddings.
+    :param delta: The base shape of the input relation embeddings.
+    :param size: The size scalar vectors of the input relation embeddings.
+    """
+    size_pos = torch.nn.functional.elu(size) + 1  # Enforce that sizes are strictly positive by passing through ELU
```

Comment:
I guess the main motivation was that the value is simply x+1 when x is positive, but since we just need the size to be positive, softplus would also work.